[Gutenberg] Add new block type badge to block inserter list

dcalhoun opened the pull request:
* **Gutenberg:** https://github.com/WordPress/gutenberg/pull/33119
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3720
* **WPiOS:** Self
* **WPAndroid:** https://github.com/wordpress-mobile/WordPress-Android/pull/15038

Fixes WordPress/gutenberg#30627. Add badge to showcase new block types. 

To test: See https://github.com/WordPress/gutenberg/pull/33119. 

## Regression Notes
1. Potential unintended areas of impact
  Modifications to the Gutenberg view and settings could cause disruption initializing the editor. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Verified the editor loaded without issue and successfully modified/saved post content. 
3. What automated tests I added (or what prevented me from doing so)
  None, as this feature is experimental at this time and any tests would be implemented in the Gutenberg editor repo.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dcalhoun committed:
Persist new block type impression count

Add Gutenberg settings field to track new block type impression counts.

dcalhoun committed:
Update gutenberg ref

dcalhoun committed:
Split tracking of editor onboarding capability and initial session

Previously, the `editorOnboarding` capability included both (1) whether
the user was within the cohort and (2) if the user had previously
launched the Gutenberg editor. The latter is utilized to only show the
initial tooltip once. However, the combination resulted in the latter
preventing the new block badges from displaying past the initial
session.

Splitting these two attributes allows for features like the new blocks
badge to remain active past the user's initial editor session.

dcalhoun committed:
Update gutenberg ref

guarani commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -941,6 +941,19 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
     }
+
+    func gutenbergDidRequestBlockTypeImpressions(_ newBlockTypes: [String]) -> [String: Int] {
+        let newImpressions = newBlockTypes.reduce(into: [:], { $0[$1] = 3 })
```
nitpick: the shorthand arguments `$0`, `$1`, etc are best avoided IMO because they reduce readability. I don't think we have any guidelines around this though.

guarani replied:
I understand this line to say, give all "new blocks" a default count of 3 "impressions". If so, it might be good to bring the `3` out into a variable such as `defaultImpressionCount`.

dcalhoun replied:
Avoiding the cryptic shorthand and unlabeled number of `3` makes total sense. However, this logic has now been moved into the JavaScript source, so the point is moot. Thank you for calling this out, though.

guarani commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -941,6 +941,19 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
+    func gutenbergDidRequestSetBlockTypeImpressionCount(_ name: String, count: Int) -> Void {
+        let newImpressions = [name: count]
+        // Merge current and new impressions favoring the new
+        gutenbergSettings.blockTypeImpressions.merge(newImpressions) { (_, new) in new }
```
Nice work with the Swift here.

guarani reviewed (commented):
👋 Great work here as usual @dcalhoun! I left some minor comments but will hold off approving since the Gutenberg side is still under review.

dcalhoun committed:
Remove logic that is now implemented in JavaScript

To de-duplicate and simplify bridging code, more logic was relocated to
the JavaScript for the React Native app.

dcalhoun committed:
Update gutenberg ref

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into add/native-new-block-badge

dcalhoun committed:
Install latest pods

guarani reviewed (approved):
As you mentioned above most of the logic that was here has moved to JavaScript, but what's left here looks great. 

I tested iOS and left an approval on the other PRs as well: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3720#pullrequestreview-720671594

dcalhoun committed:
Update gutenberg ref

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into add/native-new-block-badge

dcalhoun committed:
Update gutenberg ref to v1.59.0-alpha2 tag
